File: 0001-0100/090-subsets-ii/090-subsets-ii.js
// 90. Subsets II
// https://leetcode.com/problems/subsets-ii/description/

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function (nums) {
  function generateSubset(pos, subset) {
    res.push([...subset]);

    for (let i = pos; i < nums.length; i++) {
      if (i !== pos && nums[i] === nums[i - 1]) continue;
      subset.push(nums[i]);
      generateSubset(i + 1, subset);
      subset.pop();
    }
  }

  const res = [];
  nums.sort((a, b) => (a - b));
  generateSubset(0, []);
  return res;
};

console.info(subsetsWithDup([1, 2, 2]));

File: 0001-0100/090-subsets-ii/090-subsets-ii.cc
class Solution {
    vector<vector<int>> res;
    vector<int> path;

    void backtracking(vector<int>& nums, int start) {
        res.push_back(path);
        if (start == nums.size()) return;

        for (int i = start; i < nums.size(); ++i) {
            if (i > start && nums[i] == nums[i-1]) continue;
            path.push_back(nums[i]);
            backtracking(nums, i + 1);
            path.pop_back();
        }
    }
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        backtracking(nums, 0);
        return res;
    }
};